{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting started\n",
    "\n",
    "The basic runnable component of Pydra is a *task*. Tasks are conceptually similar to\n",
    "functions, in that they take inputs, operate on them and then return results. However,\n",
    "unlike functions, tasks are parameterised before they are executed in a separate step.\n",
    "This enables parameterised tasks to be linked together into workflows that are checked for\n",
    "errors before they are executed, and modular execution workers and environments to specified\n",
    "independently of the task being performed.\n",
    "\n",
    "Tasks can encapsulate Python functions or shell-commands, or be multi-component workflows,\n",
    "themselves constructed from task components including nested workflows.\n",
    "\n",
    "## Preparation\n",
    "\n",
    "Before we get started, lets set up some test data to play with. Here we create a sample\n",
    "JSON file in a temporary directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:18.937128Z",
     "iopub.status.busy": "2025-02-13T23:47:18.936972Z",
     "iopub.status.idle": "2025-02-13T23:47:18.940736Z",
     "shell.execute_reply": "2025-02-13T23:47:18.940343Z"
    }
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from tempfile import mkdtemp\n",
    "from pprint import pprint\n",
    "import json\n",
    "\n",
    "JSON_CONTENTS = {'a': True, 'b': 'two', 'c': 3, 'd': [7, 0.55, 6]}\n",
    "\n",
    "test_dir = Path(mkdtemp())\n",
    "json_file = test_dir / \"test.json\"\n",
    "with open(json_file, \"w\") as f:\n",
    "    json.dump(JSON_CONTENTS, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we create a directory containing 10 randomly generated [NIfTI](https://nifti.nimh.nih.gov/) files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:18.942087Z",
     "iopub.status.busy": "2025-02-13T23:47:18.941949Z",
     "iopub.status.idle": "2025-02-13T23:47:18.979505Z",
     "shell.execute_reply": "2025-02-13T23:47:18.979008Z"
    }
   },
   "outputs": [],
   "source": [
    "from fileformats.medimage import Nifti1\n",
    "\n",
    "nifti_dir = test_dir / \"nifti\"\n",
    "nifti_dir.mkdir()\n",
    "\n",
    "for i in range(10):\n",
    "    Nifti1.sample(nifti_dir, seed=i)  # Create a dummy NIfTI file in the dest. directory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that when you run concurrent processes within a Jupyter notebook the following snippet\n",
    "is also required"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:18.981229Z",
     "iopub.status.busy": "2025-02-13T23:47:18.980751Z",
     "iopub.status.idle": "2025-02-13T23:47:18.983600Z",
     "shell.execute_reply": "2025-02-13T23:47:18.983243Z"
    }
   },
   "outputs": [],
   "source": [
    "import nest_asyncio\n",
    "nest_asyncio.apply()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "## Running your first task\n",
    "\n",
    "Pre-defined task definitions are installed under the `pydra.tasks.*` namespace by separate\n",
    "task packages (e.g. `pydra-fsl`, `pydra-ants`, ...). To use a pre-defined task definition\n",
    "\n",
    "* import the class from the `pydra.tasks.*` package it is in\n",
    "* instantiate it with appropriate parameters\n",
    "* \"call\" resulting object (i.e. `my_task(...)`) to execute it as you would a function \n",
    "\n",
    "To demonstrate with an example of loading a JSON file with the\n",
    "`pydra.tasks.common.LoadJson` task, we first create an example JSON file to test with"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we can load the JSON contents back from the file using the `LoadJson` task definition\n",
    "class"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:18.984912Z",
     "iopub.status.busy": "2025-02-13T23:47:18.984775Z",
     "iopub.status.idle": "2025-02-13T23:47:19.715067Z",
     "shell.execute_reply": "2025-02-13T23:47:19.714589Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "A newer version (0.25) of nipype/pydra is available. You are using 0.25.dev216+ga56b61e5\n"
     ]
    }
   ],
   "source": [
    "# Import the task definition\n",
    "from pydra.tasks.common import LoadJson\n",
    "\n",
    "# Instantiate the task definition, providing the JSON file we want to load\n",
    "load_json = LoadJson(file=json_file)\n",
    "\n",
    "# Run the task to load the JSON file\n",
    "outputs = load_json()\n",
    "\n",
    "# Access the loaded JSON output contents and check they match original\n",
    "assert outputs.out == JSON_CONTENTS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Iterating over inputs\n",
    "\n",
    "It is straightforward to apply the same operation over a set of inputs using the `split()`\n",
    "method. For example, if we wanted to re-grid all the NIfTI images stored in a directory,\n",
    "such as the sample ones generated by the code below"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then we can by importing the `MrGrid` shell-command task from the `pydra-mrtrix3` package\n",
    "and run it over every NIfTI file in the directory using the `TaskDef.split()` method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:19.717063Z",
     "iopub.status.busy": "2025-02-13T23:47:19.716693Z",
     "iopub.status.idle": "2025-02-13T23:47:20.754510Z",
     "shell.execute_reply": "2025-02-13T23:47:20.753917Z"
    }
   },
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[5], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m mrgrid \u001b[38;5;241m=\u001b[39m MrGrid(operation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregrid\u001b[39m\u001b[38;5;124m\"\u001b[39m, voxel\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m0.5\u001b[39m,\u001b[38;5;241m0.5\u001b[39m,\u001b[38;5;241m0.5\u001b[39m))\u001b[38;5;241m.\u001b[39msplit(in_file\u001b[38;5;241m=\u001b[39mnifti_dir\u001b[38;5;241m.\u001b[39miterdir())\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# Run the task to resample all NIfTI files\u001b[39;00m\n\u001b[0;32m----> 8\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmrgrid\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m \u001b[38;5;66;03m# Print the locations of the output files\u001b[39;00m\n\u001b[1;32m     11\u001b[0m pprint(outputs\u001b[38;5;241m.\u001b[39mout_file)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:243\u001b[0m, in \u001b[0;36mTaskDef.__call__\u001b[0;34m(self, cache_dir, worker, environment, rerun, cache_locations, audit_flags, messengers, messenger_args, name, hooks, **kwargs)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Submitter(\n\u001b[1;32m    233\u001b[0m         audit_flags\u001b[38;5;241m=\u001b[39maudit_flags,\n\u001b[1;32m    234\u001b[0m         cache_dir\u001b[38;5;241m=\u001b[39mcache_dir,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    242\u001b[0m     ) \u001b[38;5;28;01mas\u001b[39;00m sub:\n\u001b[0;32m--> 243\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43msub\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    245\u001b[0m \u001b[43m            \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    246\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhooks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhooks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    247\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;66;03m# Catch any inadvertent passing of task definition parameters to the\u001b[39;00m\n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# execution call\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(e, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__notes__\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m WORKER_KWARG_FAIL_NOTE \u001b[38;5;129;01min\u001b[39;00m e\u001b[38;5;241m.\u001b[39m__notes__:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:218\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    220\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:212\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloop\u001b[38;5;241m.\u001b[39mrun_until_complete(\n\u001b[1;32m    209\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mrun_async(task, rerun\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrerun)\n\u001b[1;32m    210\u001b[0m         )\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 212\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:810\u001b[0m, in \u001b[0;36mWorkflowDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[WorkflowDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    809\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the workflow.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 810\u001b[0m     \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msubmitter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpand_workflow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:267\u001b[0m, in \u001b[0;36mSubmitter.expand_workflow\u001b[0;34m(self, workflow_task)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m tasks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m n\u001b[38;5;241m.\u001b[39mdone \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m exec_graph\u001b[38;5;241m.\u001b[39mnodes):\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m task \u001b[38;5;129;01min\u001b[39;00m tasks:\n\u001b[0;32m--> 267\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    268\u001b[0m     tasks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_runnable_tasks(exec_graph)\n\u001b[1;32m    269\u001b[0m workflow_task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mworkflow\u001b[39m\u001b[38;5;124m\"\u001b[39m: wf, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexec_graph\u001b[39m\u001b[38;5;124m\"\u001b[39m: exec_graph}\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:1010\u001b[0m, in \u001b[0;36mShellDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m   1008\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[ShellDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the shell command.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1010\u001b[0m     task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menvironment\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py:56\u001b[0m, in \u001b[0;36mNative.execute\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m     55\u001b[0m         msg \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mstdout:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(msg)\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\u001b[0mFull crash report for 'Split' task is here: /home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/workflow-cd6d45d532c73047c0fb66f9ecda6654/_error.pklz"
     ]
    }
   ],
   "source": [
    "from pydra.tasks.mrtrix3.v3_0 import MrGrid\n",
    "\n",
    "# Instantiate the task definition, \"splitting\" over all NIfTI files in the test directory\n",
    "# by splitting the \"input\" input field over all files in the directory\n",
    "mrgrid = MrGrid(operation=\"regrid\", voxel=(0.5,0.5,0.5)).split(in_file=nifti_dir.iterdir())\n",
    "\n",
    "# Run the task to resample all NIfTI files\n",
    "outputs = mrgrid()\n",
    "\n",
    "# Print the locations of the output files\n",
    "pprint(outputs.out_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is also possible to iterate over inputs in pairs/n-tuples. For example, if you wanted to use\n",
    "different voxel sizes for different images, both the list of images and the voxel sizes\n",
    "are passed to the `split()` method and their combination is specified by a tuple \"splitter\"\n",
    "\n",
    "\n",
    "Note that it is important to use a tuple not a list for the splitter definition in this\n",
    "case, because a list splitter is interpreted as the split over each combination of inputs\n",
    "(see [Splitting and combining](../explanation/splitting-combining.html) for more details\n",
    "on splitters)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:20.756166Z",
     "iopub.status.busy": "2025-02-13T23:47:20.756005Z",
     "iopub.status.idle": "2025-02-13T23:47:21.042455Z",
     "shell.execute_reply": "2025-02-13T23:47:21.041848Z"
    }
   },
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-ab511ef6bff4517710084b47bd36bae0/out_file.mif', '-voxel', '1.0,1.0,1.0']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[6], line 20\u001b[0m\n\u001b[1;32m      1\u001b[0m mrgrid_varying_vox_sizes \u001b[38;5;241m=\u001b[39m MrGrid(operation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregrid\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39msplit(\n\u001b[1;32m      2\u001b[0m     (\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124min_file\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvoxel\u001b[39m\u001b[38;5;124m\"\u001b[39m),\n\u001b[1;32m      3\u001b[0m     in_file\u001b[38;5;241m=\u001b[39mnifti_dir\u001b[38;5;241m.\u001b[39miterdir(),\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     17\u001b[0m     ],\n\u001b[1;32m     18\u001b[0m )\n\u001b[0;32m---> 20\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmrgrid_varying_vox_sizes\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     22\u001b[0m pprint(outputs\u001b[38;5;241m.\u001b[39mout_file)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:243\u001b[0m, in \u001b[0;36mTaskDef.__call__\u001b[0;34m(self, cache_dir, worker, environment, rerun, cache_locations, audit_flags, messengers, messenger_args, name, hooks, **kwargs)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Submitter(\n\u001b[1;32m    233\u001b[0m         audit_flags\u001b[38;5;241m=\u001b[39maudit_flags,\n\u001b[1;32m    234\u001b[0m         cache_dir\u001b[38;5;241m=\u001b[39mcache_dir,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    242\u001b[0m     ) \u001b[38;5;28;01mas\u001b[39;00m sub:\n\u001b[0;32m--> 243\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43msub\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    245\u001b[0m \u001b[43m            \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    246\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhooks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhooks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    247\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;66;03m# Catch any inadvertent passing of task definition parameters to the\u001b[39;00m\n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# execution call\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(e, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__notes__\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m WORKER_KWARG_FAIL_NOTE \u001b[38;5;129;01min\u001b[39;00m e\u001b[38;5;241m.\u001b[39m__notes__:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:218\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    220\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:212\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloop\u001b[38;5;241m.\u001b[39mrun_until_complete(\n\u001b[1;32m    209\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mrun_async(task, rerun\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrerun)\n\u001b[1;32m    210\u001b[0m         )\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 212\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:810\u001b[0m, in \u001b[0;36mWorkflowDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[WorkflowDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    809\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the workflow.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 810\u001b[0m     \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msubmitter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpand_workflow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:267\u001b[0m, in \u001b[0;36mSubmitter.expand_workflow\u001b[0;34m(self, workflow_task)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m tasks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m n\u001b[38;5;241m.\u001b[39mdone \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m exec_graph\u001b[38;5;241m.\u001b[39mnodes):\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m task \u001b[38;5;129;01min\u001b[39;00m tasks:\n\u001b[0;32m--> 267\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    268\u001b[0m     tasks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_runnable_tasks(exec_graph)\n\u001b[1;32m    269\u001b[0m workflow_task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mworkflow\u001b[39m\u001b[38;5;124m\"\u001b[39m: wf, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexec_graph\u001b[39m\u001b[38;5;124m\"\u001b[39m: exec_graph}\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:1010\u001b[0m, in \u001b[0;36mShellDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m   1008\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[ShellDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the shell command.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1010\u001b[0m     task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menvironment\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py:56\u001b[0m, in \u001b[0;36mNative.execute\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m     55\u001b[0m         msg \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mstdout:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(msg)\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-ab511ef6bff4517710084b47bd36bae0/out_file.mif', '-voxel', '1.0,1.0,1.0']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\u001b[0mFull crash report for 'Split' task is here: /home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/workflow-ce3d79a553a06b1cd5b23a36a4144fc8/_error.pklz"
     ]
    }
   ],
   "source": [
    "mrgrid_varying_vox_sizes = MrGrid(operation=\"regrid\").split(\n",
    "    (\"in_file\", \"voxel\"),\n",
    "    in_file=nifti_dir.iterdir(),\n",
    "    # Define a list of voxel sizes to resample the NIfTI files to,\n",
    "    # the list must be the same length as the list of NIfTI files\n",
    "    voxel=[\n",
    "        (1.0, 1.0, 1.0),\n",
    "        (1.0, 1.0, 1.0),\n",
    "        (1.0, 1.0, 1.0),\n",
    "        (0.5, 0.5, 0.5),\n",
    "        (0.75, 0.75, 0.75),\n",
    "        (0.5, 0.5, 0.5),\n",
    "        (0.5, 0.5, 0.5),\n",
    "        (1.0, 1.0, 1.0),\n",
    "        (1.25, 1.25, 1.25),\n",
    "        (1.25, 1.25, 1.25),\n",
    "    ],\n",
    ")\n",
    "\n",
    "outputs = mrgrid_varying_vox_sizes()\n",
    "\n",
    "pprint(outputs.out_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Executing tasks in parallel\n",
    "\n",
    "By default, Pydra will use the *debug* worker, which executes each task sequentially.\n",
    "This makes it easier to debug tasks and workflows, however, in most cases, once a workflow\n",
    "is tested, a concurrent worker is preferable so tasks can be executed in parallel\n",
    "(see [Workers](./3-advanced-execution.html#Workers)). To use multiple processes on a\n",
    "workstation, select the `cf` worker option when executing the task/workflow. Additional\n",
    "keyword arguments, will be passed to the worker initialisation (e.g. `n_procs=4`).\n",
    "\n",
    "Note that when multiprocessing in Python on Windows and macOS (and good practice on Linux/POSIX\n",
    "OSs for compatibility), you need to place a `if __name__ == \"__main__\"` block when\n",
    "executing in top-level scripts to allow the script to be imported, but not executed,\n",
    "by subprocesses."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:21.044085Z",
     "iopub.status.busy": "2025-02-13T23:47:21.043935Z",
     "iopub.status.idle": "2025-02-13T23:47:21.551505Z",
     "shell.execute_reply": "2025-02-13T23:47:21.550889Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-3' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-493775935d04d6631d11949224004ce7/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-493775935d04d6631d11949224004ce7/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-493775935d04d6631d11949224004ce7/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/MTX6T5bC6O3hDpHJanr4VjXi.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-5' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c0e4137c18c07e05b77dac2a94671a0a/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c0e4137c18c07e05b77dac2a94671a0a/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c0e4137c18c07e05b77dac2a94671a0a/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/XYEqaPDVxVqXTzHQlIXqXQ0u.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-4' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-2' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-7afa666d84383040618a0fa98bbf6816/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-7afa666d84383040618a0fa98bbf6816/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-7afa666d84383040618a0fa98bbf6816/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/o7hRfp9mNCCElZfobqz3xhq9.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-7' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c1b4d4812e300c92de67d72c95a847e8/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c1b4d4812e300c92de67d72c95a847e8/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c1b4d4812e300c92de67d72c95a847e8/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ogyjey4XVnHrkgn5ZYXltMT0.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-11' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-4fd1fd1c1ff5cdb044e6154ddb4739ec/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-4fd1fd1c1ff5cdb044e6154ddb4739ec/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-4fd1fd1c1ff5cdb044e6154ddb4739ec/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/Cxi1aF3fIMcxRCSjogF5KWxU.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-9' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-498e9e06b66bfc6e8acb06611ac1bb3c/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-498e9e06b66bfc6e8acb06611ac1bb3c/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-498e9e06b66bfc6e8acb06611ac1bb3c/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/76dfZTPtLLKjAyS96HBqcbCt.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-8' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-fdc6a8c6810dd12f574f0acd72993235/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-fdc6a8c6810dd12f574f0acd72993235/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-fdc6a8c6810dd12f574f0acd72993235/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/0UAqFzWsDK4FrUMp48Y3tT3Q.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-10' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-9ac88bc2b1899b78ee8d6c90bb226a6e/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-9ac88bc2b1899b78ee8d6c90bb226a6e/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-9ac88bc2b1899b78ee8d6c90bb226a6e/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/oHwLMeaZqo9DZDNjN1GTPdVK.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Task exception was never retrieved\n",
      "future: <Task finished name='Task-6' coro=<ConcurrentFuturesWorker.run() done, defined at /usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:189> exception=RuntimeError('Error running \\'MrGrid\\' task with [\\'mrgrid\\', \\'/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\\', \\'regrid\\', \\'/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-0006560da876a855fb525c2b5595ef91/out_file.mif\\', \\'-voxel\\', \\'0.5,0.5,0.5\\']:\\n\\nstderr:\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\\x1b[0m\\nmrgrid: \\x1b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\\x1b[0m\\n')>\n",
      "concurrent.futures.process._RemoteTraceback: \n",
      "\"\"\"\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/concurrent/futures/process.py\", line 264, in _process_worker\n",
      "    r = call_item.fn(*call_item.args, **call_item.kwargs)\n",
      "        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 205, in unpickle_and_run\n",
      "    return task.run(rerun=rerun)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py\", line 367, in run\n",
      "    self.definition._run(self)\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py\", line 1010, in _run\n",
      "    task.return_values = task.environment.execute(task)\n",
      "                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py\", line 56, in execute\n",
      "    raise RuntimeError(msg)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-0006560da876a855fb525c2b5595ef91/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\u001b[0m\n",
      "\n",
      "\"\"\"\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 316, in __step_run_and_handle_result\n",
      "    result = coro.throw(exc)\n",
      "             ^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py\", line 197, in run\n",
      "    return await self.loop.run_in_executor(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 289, in __await__\n",
      "    yield self  # This tells Task to wait for completion.\n",
      "    ^^^^^^^^^^\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/tasks.py\", line 385, in __wakeup\n",
      "    future.result()\n",
      "  File \"/usr/share/miniconda/lib/python3.12/asyncio/futures.py\", line 202, in result\n",
      "    raise self._exception.with_traceback(self._exception_tb)\n",
      "RuntimeError: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii', 'regrid', '/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-0006560da876a855fb525c2b5595ef91/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n",
      "\n",
      "stderr:\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\u001b[0m\n",
      "mrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/ujOeHwdFcAefAZhnM6Jy8c1r.nii\"\u001b[0m\n",
      "\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Task 'MrGrid' raised an error, full crash report is here: \n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-0006560da876a855fb525c2b5595ef91/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c1b4d4812e300c92de67d72c95a847e8/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-9ac88bc2b1899b78ee8d6c90bb226a6e/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-493775935d04d6631d11949224004ce7/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-fdc6a8c6810dd12f574f0acd72993235/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-4fd1fd1c1ff5cdb044e6154ddb4739ec/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c0e4137c18c07e05b77dac2a94671a0a/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-498e9e06b66bfc6e8acb06611ac1bb3c/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-7afa666d84383040618a0fa98bbf6816/_error.pklz",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:771\u001b[0m, in \u001b[0;36mWorkflowOutputs._from_task\u001b[0;34m(cls, task)\u001b[0m\n\u001b[1;32m    770\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 771\u001b[0m     val_out \u001b[38;5;241m=\u001b[39m \u001b[43mlazy_field\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_get_value\u001b[49m\u001b[43m(\u001b[49m\u001b[43mexec_graph\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    772\u001b[0m     output_wf[name] \u001b[38;5;241m=\u001b[39m val_out\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/lazy.py:171\u001b[0m, in \u001b[0;36mLazyOutField._get_value\u001b[0;34m(self, graph, state_index)\u001b[0m\n\u001b[1;32m    169\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m val\n\u001b[0;32m--> 171\u001b[0m value \u001b[38;5;241m=\u001b[39m \u001b[43mget_nested\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msplit_depth\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    172\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m value\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/lazy.py:138\u001b[0m, in \u001b[0;36mLazyOutField._get_value.<locals>.get_nested\u001b[0;34m(task, depth)\u001b[0m\n\u001b[1;32m    137\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(task, StateArray):\n\u001b[0;32m--> 138\u001b[0m     val \u001b[38;5;241m=\u001b[39m [\u001b[43mget_nested\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mt\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdepth\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdepth\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m-\u001b[39;49m\u001b[43m \u001b[49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m)\u001b[49m \u001b[38;5;28;01mfor\u001b[39;00m t \u001b[38;5;129;01min\u001b[39;00m task]\n\u001b[1;32m    139\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m depth:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/lazy.py:143\u001b[0m, in \u001b[0;36mLazyOutField._get_value.<locals>.get_nested\u001b[0;34m(task, depth)\u001b[0m\n\u001b[1;32m    142\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m task\u001b[38;5;241m.\u001b[39merrored:\n\u001b[0;32m--> 143\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    144\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCannot retrieve value for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_field\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m from \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_node\u001b[38;5;241m.\u001b[39mname\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m as \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    145\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthe node errored\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    146\u001b[0m     )\n\u001b[1;32m    147\u001b[0m res \u001b[38;5;241m=\u001b[39m task\u001b[38;5;241m.\u001b[39mresult()\n",
      "\u001b[0;31mValueError\u001b[0m: Cannot retrieve value for out_file from MrGrid as the node errored",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[7], line 5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;18m__name__\u001b[39m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__main__\u001b[39m\u001b[38;5;124m\"\u001b[39m:  \u001b[38;5;66;03m# <-- Add this block to allow the script to imported by subprocesses\u001b[39;00m\n\u001b[1;32m      4\u001b[0m     mrgrid \u001b[38;5;241m=\u001b[39m MrGrid(operation\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mregrid\u001b[39m\u001b[38;5;124m\"\u001b[39m, voxel\u001b[38;5;241m=\u001b[39m(\u001b[38;5;241m0.5\u001b[39m,\u001b[38;5;241m0.5\u001b[39m,\u001b[38;5;241m0.5\u001b[39m))\u001b[38;5;241m.\u001b[39msplit(in_file\u001b[38;5;241m=\u001b[39mnifti_dir\u001b[38;5;241m.\u001b[39miterdir())\n\u001b[0;32m----> 5\u001b[0m     outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmrgrid\u001b[49m\u001b[43m(\u001b[49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mcf\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mn_procs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m4\u001b[39;49m\u001b[43m)\u001b[49m  \u001b[38;5;66;03m# <-- Select the \"cf\" worker here\u001b[39;00m\n\u001b[1;32m      6\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mstr\u001b[39m(p) \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m outputs\u001b[38;5;241m.\u001b[39mout_file))\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:243\u001b[0m, in \u001b[0;36mTaskDef.__call__\u001b[0;34m(self, cache_dir, worker, environment, rerun, cache_locations, audit_flags, messengers, messenger_args, name, hooks, **kwargs)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Submitter(\n\u001b[1;32m    233\u001b[0m         audit_flags\u001b[38;5;241m=\u001b[39maudit_flags,\n\u001b[1;32m    234\u001b[0m         cache_dir\u001b[38;5;241m=\u001b[39mcache_dir,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    242\u001b[0m     ) \u001b[38;5;28;01mas\u001b[39;00m sub:\n\u001b[0;32m--> 243\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43msub\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    245\u001b[0m \u001b[43m            \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    246\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhooks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhooks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    247\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;66;03m# Catch any inadvertent passing of task definition parameters to the\u001b[39;00m\n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# execution call\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(e, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__notes__\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m WORKER_KWARG_FAIL_NOTE \u001b[38;5;129;01min\u001b[39;00m e\u001b[38;5;241m.\u001b[39m__notes__:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:218\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    220\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:208\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    206\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m datetime\u001b[38;5;241m.\u001b[39mnow()\n\u001b[1;32m    207\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mis_async:  \u001b[38;5;66;03m# Only workflow tasks can be async\u001b[39;00m\n\u001b[0;32m--> 208\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mloop\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_until_complete\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    209\u001b[0m \u001b[43m        \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun_async\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    210\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    211\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    212\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mrun(task, rerun\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrerun)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/nest_asyncio.py:98\u001b[0m, in \u001b[0;36m_patch_loop.<locals>.run_until_complete\u001b[0;34m(self, future)\u001b[0m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m f\u001b[38;5;241m.\u001b[39mdone():\n\u001b[1;32m     96\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\n\u001b[1;32m     97\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEvent loop stopped before Future completed.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m---> 98\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mf\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mresult\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/asyncio/futures.py:202\u001b[0m, in \u001b[0;36mFuture.result\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    200\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m__log_traceback \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m    201\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 202\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception\u001b[38;5;241m.\u001b[39mwith_traceback(\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_exception_tb)\n\u001b[1;32m    203\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_result\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/asyncio/tasks.py:314\u001b[0m, in \u001b[0;36mTask.__step_run_and_handle_result\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    310\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    311\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m exc \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    312\u001b[0m         \u001b[38;5;66;03m# We use the `send` method directly, because coroutines\u001b[39;00m\n\u001b[1;32m    313\u001b[0m         \u001b[38;5;66;03m# don't have `__iter__` and `__next__` methods.\u001b[39;00m\n\u001b[0;32m--> 314\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43mcoro\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mNone\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m    315\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    316\u001b[0m         result \u001b[38;5;241m=\u001b[39m coro\u001b[38;5;241m.\u001b[39mthrow(exc)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:52\u001b[0m, in \u001b[0;36mWorker.run_async\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;28;01masync\u001b[39;00m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun_async\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m, rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     51\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m task\u001b[38;5;241m.\u001b[39mis_async:\n\u001b[0;32m---> 52\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mawait\u001b[39;00m task\u001b[38;5;241m.\u001b[39mrun_async(rerun\u001b[38;5;241m=\u001b[39mrerun)\n\u001b[1;32m     53\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m     54\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m task\u001b[38;5;241m.\u001b[39mrun(rerun\u001b[38;5;241m=\u001b[39mrerun)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:422\u001b[0m, in \u001b[0;36mTask.run_async\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    420\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[1;32m    421\u001b[0m     \u001b[38;5;28;01mawait\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39m_run_async(\u001b[38;5;28mself\u001b[39m)\n\u001b[0;32m--> 422\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mOutputs\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_from_task\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    423\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m    424\u001b[0m     etype, \u001b[38;5;28meval\u001b[39m, etr \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mexc_info()\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:784\u001b[0m, in \u001b[0;36mWorkflowOutputs._from_task\u001b[0;34m(cls, task)\u001b[0m\n\u001b[1;32m    782\u001b[0m             \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m err_files:\n\u001b[1;32m    783\u001b[0m                 \u001b[38;5;28;01mraise\u001b[39;00m\n\u001b[0;32m--> 784\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m    785\u001b[0m                 \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mlazy_field\u001b[38;5;241m.\u001b[39m_node\u001b[38;5;241m.\u001b[39mname\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m raised an error, full crash report is \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    786\u001b[0m                 \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhere: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    787\u001b[0m                 \u001b[38;5;241m+\u001b[39m (\n\u001b[1;32m    788\u001b[0m                     \u001b[38;5;28mstr\u001b[39m(err_files[\u001b[38;5;241m0\u001b[39m])\n\u001b[1;32m    789\u001b[0m                     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(err_files) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m    790\u001b[0m                     \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(\u001b[38;5;28mstr\u001b[39m(f) \u001b[38;5;28;01mfor\u001b[39;00m f \u001b[38;5;129;01min\u001b[39;00m err_files)\n\u001b[1;32m    791\u001b[0m                 )\n\u001b[1;32m    792\u001b[0m             )\n\u001b[1;32m    793\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m attrs\u001b[38;5;241m.\u001b[39mevolve(outputs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39moutput_wf)\n",
      "\u001b[0;31mValueError\u001b[0m: Task 'MrGrid' raised an error, full crash report is here: \n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-0006560da876a855fb525c2b5595ef91/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c1b4d4812e300c92de67d72c95a847e8/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-9ac88bc2b1899b78ee8d6c90bb226a6e/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-493775935d04d6631d11949224004ce7/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-fdc6a8c6810dd12f574f0acd72993235/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-4fd1fd1c1ff5cdb044e6154ddb4739ec/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-c0e4137c18c07e05b77dac2a94671a0a/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-498e9e06b66bfc6e8acb06611ac1bb3c/_error.pklz\n/home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/shell-7afa666d84383040618a0fa98bbf6816/_error.pklz",
      "\u001b[0mFull crash report for 'Split' task is here: /home/runner/.cache/pydra/0.25.dev216+ga56b61e5/run-cache/workflow-cd6d45d532c73047c0fb66f9ecda6654/_error.pklz"
     ]
    }
   ],
   "source": [
    "from pydra.tasks.mrtrix3.v3_0 import MrGrid\n",
    "\n",
    "if __name__ == \"__main__\":  # <-- Add this block to allow the script to imported by subprocesses\n",
    "    mrgrid = MrGrid(operation=\"regrid\", voxel=(0.5,0.5,0.5)).split(in_file=nifti_dir.iterdir())\n",
    "    outputs = mrgrid(worker=\"cf\", n_procs=4)  # <-- Select the \"cf\" worker here\n",
    "    print(\"\\n\".join(str(p) for p in outputs.out_file))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## File-system locations\n",
    "\n",
    "Output and intermediate files are typically generated during the course of a workflow/task run.\n",
    "In addition to this, Pydra generates a cache directory for each task, in which\n",
    "the task definition, results and any errors are stored in [cloudpickle](https://github.com/cloudpipe/cloudpickle)\n",
    "files for future reference (see [Troubleshooting](./troubleshooting.html)).\n",
    "By default, these cache directories are stored in a platform-specific application-cache\n",
    "directory\n",
    "\n",
    "* Windows: `C:\\Users\\<username>\\AppData\\Local\\pydra\\<pydra-version>\\run-cache`\n",
    "* Linux: `/home/<username>/.cache/pydra/<pydra-version>/run-cache`\n",
    "* macOS: `/Users/<username>/Library/Caches/pydra/<pydra-version>/run-cache`\n",
    "\n",
    "When a task runs, a unique hash is generated by the combination of all the inputs to the\n",
    "task and the operation to be performed. This hash is used to name the task cache directory\n",
    "within the specified cache root. Therefore, if you use the same cache\n",
    "root and in a subsequent run the same task is executed with the same\n",
    "inputs, then the path of its cache directory will be the same, and if Pydra finds\n",
    "existing results at that path, then the outputs generated by the previous run will be\n",
    "reused.\n",
    "\n",
    "This cache will grow as more runs are called, therefore care needs to be taken to ensure\n",
    "there is enough space on the target disk. Since the cache will be constantly To specify\n",
    "a different location for this cache, simply provide the `cache_root` keyword argument to the execution call"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:21.553394Z",
     "iopub.status.busy": "2025-02-13T23:47:21.553113Z",
     "iopub.status.idle": "2025-02-13T23:47:21.777568Z",
     "shell.execute_reply": "2025-02-13T23:47:21.777030Z"
    }
   },
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/pydra-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[8], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmrgrid\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcache_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mPath\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m~/pydra-cache\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpanduser\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      3\u001b[0m pprint(outputs)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:243\u001b[0m, in \u001b[0;36mTaskDef.__call__\u001b[0;34m(self, cache_dir, worker, environment, rerun, cache_locations, audit_flags, messengers, messenger_args, name, hooks, **kwargs)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Submitter(\n\u001b[1;32m    233\u001b[0m         audit_flags\u001b[38;5;241m=\u001b[39maudit_flags,\n\u001b[1;32m    234\u001b[0m         cache_dir\u001b[38;5;241m=\u001b[39mcache_dir,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    242\u001b[0m     ) \u001b[38;5;28;01mas\u001b[39;00m sub:\n\u001b[0;32m--> 243\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43msub\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    245\u001b[0m \u001b[43m            \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    246\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhooks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhooks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    247\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;66;03m# Catch any inadvertent passing of task definition parameters to the\u001b[39;00m\n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# execution call\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(e, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__notes__\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m WORKER_KWARG_FAIL_NOTE \u001b[38;5;129;01min\u001b[39;00m e\u001b[38;5;241m.\u001b[39m__notes__:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:218\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    220\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:212\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloop\u001b[38;5;241m.\u001b[39mrun_until_complete(\n\u001b[1;32m    209\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mrun_async(task, rerun\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrerun)\n\u001b[1;32m    210\u001b[0m         )\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 212\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:810\u001b[0m, in \u001b[0;36mWorkflowDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[WorkflowDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    809\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the workflow.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 810\u001b[0m     \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msubmitter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpand_workflow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:267\u001b[0m, in \u001b[0;36mSubmitter.expand_workflow\u001b[0;34m(self, workflow_task)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m tasks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m n\u001b[38;5;241m.\u001b[39mdone \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m exec_graph\u001b[38;5;241m.\u001b[39mnodes):\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m task \u001b[38;5;129;01min\u001b[39;00m tasks:\n\u001b[0;32m--> 267\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    268\u001b[0m     tasks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_runnable_tasks(exec_graph)\n\u001b[1;32m    269\u001b[0m workflow_task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mworkflow\u001b[39m\u001b[38;5;124m\"\u001b[39m: wf, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexec_graph\u001b[39m\u001b[38;5;124m\"\u001b[39m: exec_graph}\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:1010\u001b[0m, in \u001b[0;36mShellDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m   1008\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[ShellDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the shell command.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1010\u001b[0m     task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menvironment\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py:56\u001b[0m, in \u001b[0;36mNative.execute\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m     55\u001b[0m         msg \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mstdout:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(msg)\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/pydra-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\u001b[0mFull crash report for 'Split' task is here: /home/runner/pydra-cache/workflow-cd6d45d532c73047c0fb66f9ecda6654/_error.pklz"
     ]
    }
   ],
   "source": [
    "outputs = mrgrid(cache_dir=Path(\"~/pydra-cache\").expanduser())\n",
    "\n",
    "pprint(outputs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To check alternative cache roots, while storing any generated task cache dirs in the \n",
    "specified cache root, the `cache_locations` keyword argument can be used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2025-02-13T23:47:21.779345Z",
     "iopub.status.busy": "2025-02-13T23:47:21.779023Z",
     "iopub.status.idle": "2025-02-13T23:47:22.007076Z",
     "shell.execute_reply": "2025-02-13T23:47:22.006490Z"
    }
   },
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/new-pydra-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m my_cache_dir \u001b[38;5;241m=\u001b[39m Path(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m~/new-pydra-cache\u001b[39m\u001b[38;5;124m\"\u001b[39m)\u001b[38;5;241m.\u001b[39mexpanduser()\n\u001b[1;32m      4\u001b[0m my_cache_dir\u001b[38;5;241m.\u001b[39mmkdir(exist_ok\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m----> 6\u001b[0m outputs \u001b[38;5;241m=\u001b[39m \u001b[43mmrgrid\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmy_cache_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m      8\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcache_locations\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43m[\u001b[49m\u001b[43mdefault_run_cache_dir\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m      9\u001b[0m \u001b[43m)\u001b[49m\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28mprint\u001b[39m(outputs)\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:243\u001b[0m, in \u001b[0;36mTaskDef.__call__\u001b[0;34m(self, cache_dir, worker, environment, rerun, cache_locations, audit_flags, messengers, messenger_args, name, hooks, **kwargs)\u001b[0m\n\u001b[1;32m    231\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    232\u001b[0m     \u001b[38;5;28;01mwith\u001b[39;00m Submitter(\n\u001b[1;32m    233\u001b[0m         audit_flags\u001b[38;5;241m=\u001b[39maudit_flags,\n\u001b[1;32m    234\u001b[0m         cache_dir\u001b[38;5;241m=\u001b[39mcache_dir,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    241\u001b[0m         \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs,\n\u001b[1;32m    242\u001b[0m     ) \u001b[38;5;28;01mas\u001b[39;00m sub:\n\u001b[0;32m--> 243\u001b[0m         result \u001b[38;5;241m=\u001b[39m \u001b[43msub\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    244\u001b[0m \u001b[43m            \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    245\u001b[0m \u001b[43m            \u001b[49m\u001b[43mname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    246\u001b[0m \u001b[43m            \u001b[49m\u001b[43mhooks\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mhooks\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    247\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    248\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    249\u001b[0m     \u001b[38;5;66;03m# Catch any inadvertent passing of task definition parameters to the\u001b[39;00m\n\u001b[1;32m    250\u001b[0m     \u001b[38;5;66;03m# execution call\u001b[39;00m\n\u001b[1;32m    251\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(e, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m__notes__\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;129;01mand\u001b[39;00m WORKER_KWARG_FAIL_NOTE \u001b[38;5;129;01min\u001b[39;00m e\u001b[38;5;241m.\u001b[39m__notes__:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:218\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n\u001b[0;32m--> 218\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m e\n\u001b[1;32m    219\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m    220\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrun_start_time \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:212\u001b[0m, in \u001b[0;36mSubmitter.__call__\u001b[0;34m(self, task_def, name, hooks)\u001b[0m\n\u001b[1;32m    208\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mloop\u001b[38;5;241m.\u001b[39mrun_until_complete(\n\u001b[1;32m    209\u001b[0m             \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mworker\u001b[38;5;241m.\u001b[39mrun_async(task, rerun\u001b[38;5;241m=\u001b[39m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mrerun)\n\u001b[1;32m    210\u001b[0m         )\n\u001b[1;32m    211\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 212\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    213\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    214\u001b[0m     e\u001b[38;5;241m.\u001b[39madd_note(\n\u001b[1;32m    215\u001b[0m         \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mFull crash report for \u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mtype\u001b[39m(task_def)\u001b[38;5;241m.\u001b[39m\u001b[38;5;18m__name__\u001b[39m\u001b[38;5;132;01m!r}\u001b[39;00m\u001b[38;5;124m task is here: \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    216\u001b[0m         \u001b[38;5;241m+\u001b[39m \u001b[38;5;28mstr\u001b[39m(task\u001b[38;5;241m.\u001b[39moutput_dir \u001b[38;5;241m/\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m_error.pklz\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    217\u001b[0m     )\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:810\u001b[0m, in \u001b[0;36mWorkflowDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m    808\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[WorkflowDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    809\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the workflow.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 810\u001b[0m     \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msubmitter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexpand_workflow\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/submitter.py:267\u001b[0m, in \u001b[0;36mSubmitter.expand_workflow\u001b[0;34m(self, workflow_task)\u001b[0m\n\u001b[1;32m    265\u001b[0m \u001b[38;5;28;01mwhile\u001b[39;00m tasks \u001b[38;5;129;01mor\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;129;01mnot\u001b[39;00m n\u001b[38;5;241m.\u001b[39mdone \u001b[38;5;28;01mfor\u001b[39;00m n \u001b[38;5;129;01min\u001b[39;00m exec_graph\u001b[38;5;241m.\u001b[39mnodes):\n\u001b[1;32m    266\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m task \u001b[38;5;129;01min\u001b[39;00m tasks:\n\u001b[0;32m--> 267\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mworker\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    268\u001b[0m     tasks \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_runnable_tasks(exec_graph)\n\u001b[1;32m    269\u001b[0m workflow_task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mworkflow\u001b[39m\u001b[38;5;124m\"\u001b[39m: wf, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mexec_graph\u001b[39m\u001b[38;5;124m\"\u001b[39m: exec_graph}\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/workers.py:163\u001b[0m, in \u001b[0;36mDebugWorker.run\u001b[0;34m(self, task, rerun)\u001b[0m\n\u001b[1;32m    157\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mrun\u001b[39m(\n\u001b[1;32m    158\u001b[0m     \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    159\u001b[0m     task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[DefType]\u001b[39m\u001b[38;5;124m\"\u001b[39m,\n\u001b[1;32m    160\u001b[0m     rerun: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    161\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mResult\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m    162\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run a task.\"\"\"\u001b[39;00m\n\u001b[0;32m--> 163\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrerun\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrerun\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/core.py:367\u001b[0m, in \u001b[0;36mTask.run\u001b[0;34m(self, rerun)\u001b[0m\n\u001b[1;32m    365\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    366\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39maudit\u001b[38;5;241m.\u001b[39mmonitor()\n\u001b[0;32m--> 367\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdefinition\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_run\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    368\u001b[0m     result\u001b[38;5;241m.\u001b[39moutputs \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdefinition\u001b[38;5;241m.\u001b[39mOutputs\u001b[38;5;241m.\u001b[39m_from_task(\u001b[38;5;28mself\u001b[39m)\n\u001b[1;32m    369\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/specs.py:1010\u001b[0m, in \u001b[0;36mShellDef._run\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m   1008\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21m_run\u001b[39m(\u001b[38;5;28mself\u001b[39m, task: \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTask[ShellDef]\u001b[39m\u001b[38;5;124m\"\u001b[39m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1009\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m\"\"\"Run the shell command.\"\"\"\u001b[39;00m\n\u001b[0;32m-> 1010\u001b[0m     task\u001b[38;5;241m.\u001b[39mreturn_values \u001b[38;5;241m=\u001b[39m \u001b[43mtask\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43menvironment\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mexecute\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtask\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/usr/share/miniconda/lib/python3.12/site-packages/pydra/engine/environments.py:56\u001b[0m, in \u001b[0;36mNative.execute\u001b[0;34m(self, task)\u001b[0m\n\u001b[1;32m     54\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]:\n\u001b[1;32m     55\u001b[0m         msg \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124mstdout:\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m+\u001b[39m output[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mstdout\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[0;32m---> 56\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(msg)\n\u001b[1;32m     57\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m output\n",
      "\u001b[0;31mRuntimeError\u001b[0m: Error running 'MrGrid' task with ['mrgrid', '/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii', 'regrid', '/home/runner/new-pydra-cache/shell-76ca6bb1847e32a4cc9dc37c609baa0d/out_file.mif', '-voxel', '0.5,0.5,0.5']:\n\nstderr:\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-1.1 format (sizeof_hdr != 348)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\" is not in NIFTI-2 format (sizeof_hdr != 540)\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] unknown format for image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\nmrgrid: \u001b[01;31m[ERROR] error opening image \"/tmp/tmpt9an3fws/nifti/i0VpEBOWfbZAVaBSo63bbH6x.nii\"\u001b[0m\n",
      "\u001b[0mFull crash report for 'Split' task is here: /home/runner/new-pydra-cache/workflow-cd6d45d532c73047c0fb66f9ecda6654/_error.pklz"
     ]
    }
   ],
   "source": [
    "from pydra.utils import default_run_cache_dir\n",
    "\n",
    "my_cache_dir = Path(\"~/new-pydra-cache\").expanduser()\n",
    "my_cache_dir.mkdir(exist_ok=True)\n",
    "\n",
    "outputs = mrgrid(\n",
    "    cache_dir=my_cache_dir,\n",
    "    cache_locations=[default_run_cache_dir]\n",
    ")\n",
    "\n",
    "print(outputs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "wf12",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
